network/simple-onchain-discovery/src/lib.rs

```diff
@@ -45,95 +50,136 @@ pub static EVENT_PROCESSING_LOOP_BUSY_DURATION_S: Lazy<DurationHistogram> = Lazy
     ReconfigSubscription::subscribe_all(ON_CHAIN_CONFIG_REGISTRY.to_vec(), vec![])
 }
 
-/// Extract the network_address from the provided config, depending on role.
-fn network_address(role: RoleType, config: &ValidatorConfig) -> Result<NetworkAddress, lcs::Error> {
-    match role {
-        RoleType::Validator => NetworkAddress::try_from(&config.validator_network_address),
-        RoleType::FullNode => NetworkAddress::try_from(&config.full_node_network_address),
-    }
-}
-
-/// Extracts the public key from the provided config, depending on role.
-fn public_key(role: RoleType, config: &ValidatorConfig) -> x25519::PublicKey {
-    match role {
-        RoleType::Validator => config.validator_network_identity_public_key,
-        RoleType::FullNode => config.full_node_network_identity_public_key,
-    }
+fn decrypt_validator_netaddr(
+    root_key_map: &HashMap<RootKeyVersion, RootKey>,
+    account: &AccountAddress,
+    addr_idx: u32,
+    raw_enc_addr: RawEncNetworkAddress,
+) -> Result<RawNetworkAddress> {
+    let enc_addr = EncNetworkAddress::try_from(&raw_enc_addr)
+        .map_err(anyhow::Error::new)
+        .with_context(|| {
+            format_err!(
+                "error deserializing encrypted network address: {:?}",
+                raw_enc_addr
+            )
+        })?;
+
+    let key_version = enc_addr.root_key_version();
+    let root_key = root_key_map
+        .get(&key_version)
+        .ok_or_else(|| format_err!("no root_key for version: {}", key_version))?;
+    let raw_addr = enc_addr.decrypt(root_key, account, addr_idx)?;
+    Ok(raw_addr)
 }
 
 /// Extracts a set of ConnectivityRequests from a ValidatorSet which are appropriate for a network with type role.
-fn extract_updates(role: RoleType, node_set: ValidatorSet) -> Vec<ConnectivityRequest> {
-    let node_list = node_set.payload().to_vec();
-
-    let mut updates = Vec::new();
+fn extract_updates(
+    role: RoleType,
+    root_key_map: &HashMap<RootKeyVersion, RootKey>,
+    node_set: ValidatorSet,
+) -> Vec<ConnectivityRequest> {
+    // TODO(philiphayes): remove this after removing explicit pubkey field
+    let explicit_pubkeys: HashMap<_, _> = node_set
+        .payload()
+        .iter()
+        .map(|info| {
+            let peer_id = *info.account_address();
+            let pubkey = match role {
+                RoleType::Validator => info.config().validator_network_identity_public_key,
+                RoleType::FullNode => info.config().full_node_network_identity_public_key,
+            };
+            (peer_id, pubkey)
+        })
+        .collect();
 
     // Collect the set of address updates.
-    let address_map = node_list
+    let new_peer_addrs: HashMap<_, _> = node_set
+        .into_iter()
+        .map(|info| {
+            let peer_id = *info.account_address();
+            let config = info.into_config();
+            // only one field currently, though this will change soon
+            let addr_idx = 0;
```

Comment:
Is this `addr_idx` per peer?

Reply:
yeah, this is in preparation for

```rust
struct ValidatorConfig {
    // ..
    validator_network_addresses: Vec<Vec<u8>>,
}
```

i.e., there are multiple advertised addresses per validator (and full node)

so the addr_idx is the index of the address in this list of addresses